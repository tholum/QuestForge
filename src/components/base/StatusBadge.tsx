import * as React from "react"
import { cva, type VariantProps } from "class-variance-authority"
import { 
  CheckCircle2, 
  Clock, 
  AlertCircle, 
  XCircle, 
  Pause, 
  Play, 
  Target,
  Zap,
  Calendar
} from "lucide-react"

import { cn } from "@/lib/utils"
import { Badge } from "@/components/ui/badge"

const statusBadgeVariants = cva(
  "inline-flex items-center gap-1.5 font-medium transition-all duration-200",
  {
    variants: {
      status: {
        // Goal statuses
        active: "bg-blue-100 text-blue-800 border-blue-200 dark:bg-blue-900/30 dark:text-blue-300 dark:border-blue-800",
        completed: "bg-green-100 text-green-800 border-green-200 dark:bg-green-900/30 dark:text-green-300 dark:border-green-800",
        paused: "bg-yellow-100 text-yellow-800 border-yellow-200 dark:bg-yellow-900/30 dark:text-yellow-300 dark:border-yellow-800",
        cancelled: "bg-red-100 text-red-800 border-red-200 dark:bg-red-900/30 dark:text-red-300 dark:border-red-800",
        pending: "bg-gray-100 text-gray-800 border-gray-200 dark:bg-gray-900/30 dark:text-gray-300 dark:border-gray-800",
        overdue: "bg-red-100 text-red-800 border-red-200 animate-pulse dark:bg-red-900/30 dark:text-red-300 dark:border-red-800",
        
        // Priority levels
        low: "bg-green-100 text-green-700 border-green-200 dark:bg-green-900/20 dark:text-green-400 dark:border-green-800",
        medium: "bg-yellow-100 text-yellow-700 border-yellow-200 dark:bg-yellow-900/20 dark:text-yellow-400 dark:border-yellow-800",
        high: "bg-orange-100 text-orange-700 border-orange-200 dark:bg-orange-900/20 dark:text-orange-400 dark:border-orange-800",
        critical: "bg-red-100 text-red-700 border-red-200 dark:bg-red-900/20 dark:text-red-400 dark:border-red-800",
        
        // Progress states
        notStarted: "bg-gray-100 text-gray-600 border-gray-200 dark:bg-gray-900/20 dark:text-gray-400 dark:border-gray-800",
        inProgress: "bg-blue-100 text-blue-700 border-blue-200 dark:bg-blue-900/20 dark:text-blue-400 dark:border-blue-800",
        review: "bg-purple-100 text-purple-700 border-purple-200 dark:bg-purple-900/20 dark:text-purple-400 dark:border-purple-800",
        
        // Gamification statuses
        streak: "bg-orange-100 text-orange-700 border-orange-200 dark:bg-orange-900/20 dark:text-orange-400 dark:border-orange-800",
        achievement: "bg-yellow-100 text-yellow-700 border-yellow-200 dark:bg-yellow-900/20 dark:text-yellow-400 dark:border-yellow-800",
        levelUp: "bg-purple-100 text-purple-700 border-purple-200 animate-pulse dark:bg-purple-900/20 dark:text-purple-400 dark:border-purple-800",
      },
      size: {
        xs: "text-xs px-1.5 py-0.5 min-h-5",
        sm: "text-xs px-2 py-1 min-h-6",
        default: "text-sm px-2.5 py-1 min-h-7",
        lg: "text-sm px-3 py-1.5 min-h-8",
      },
      variant: {
        default: "border",
        solid: "border-transparent",
        outline: "bg-transparent border",
        ghost: "bg-transparent border-transparent",
      },
      rounded: {
        default: "rounded-md",
        sm: "rounded-sm",
        lg: "rounded-lg",
        full: "rounded-full",
        none: "rounded-none",
      },
    },
    defaultVariants: {
      status: "active",
      size: "default",
      variant: "default",
      rounded: "default",
    },
  }
)

const statusIcons = {
  // Goal statuses
  active: Play,
  completed: CheckCircle2,
  paused: Pause,
  cancelled: XCircle,
  pending: Clock,
  overdue: AlertCircle,
  
  // Priority levels
  low: Target,
  medium: Target,
  high: Target,
  critical: AlertCircle,
  
  // Progress states
  notStarted: Clock,
  inProgress: Play,
  review: AlertCircle,
  
  // Gamification
  streak: Zap,
  achievement: Target,
  levelUp: Zap,
} as const

const statusLabels = {
  // Goal statuses
  active: "Active",
  completed: "Completed",
  paused: "Paused",
  cancelled: "Cancelled",
  pending: "Pending",
  overdue: "Overdue",
  
  // Priority levels
  low: "Low Priority",
  medium: "Medium Priority",
  high: "High Priority",
  critical: "Critical",
  
  // Progress states
  notStarted: "Not Started",
  inProgress: "In Progress",
  review: "In Review",
  
  // Gamification
  streak: "Streak",
  achievement: "Achievement",
  levelUp: "Level Up!",
} as const

export interface IStatusBadgeProps extends
  React.HTMLAttributes<HTMLDivElement>,
  VariantProps<typeof statusBadgeVariants> {
  customLabel?: string
  showIcon?: boolean
  customIcon?: React.ReactNode
  pulse?: boolean
  count?: number
  tooltip?: string
}

const StatusBadge = React.forwardRef<HTMLDivElement, IStatusBadgeProps>(
  ({
    className,
    status = "active",
    size,
    variant,
    rounded,
    customLabel,
    showIcon = true,
    customIcon,
    pulse = false,
    count,
    tooltip,
    ...props
  }, ref) => {
    const IconComponent = statusIcons[status as keyof typeof statusIcons]
    const defaultLabel = statusLabels[status as keyof typeof statusLabels]
    const label = customLabel || defaultLabel

    const badgeContent = (
      <>
        {showIcon && (customIcon || (IconComponent && <IconComponent className="w-3 h-3" />))}
        <span className="truncate">
          {label}
          {count !== undefined && ` (${count})`}
        </span>
      </>
    )

    return (
      <Badge
        ref={ref}
        variant="outline"
        title={tooltip}
        className={cn(
          statusBadgeVariants({ status, size, variant, rounded }),
          pulse && "animate-pulse",
          status === "overdue" && "animate-pulse",
          status === "levelUp" && "animate-pulse",
          className
        )}
        {...props}
      >
        {badgeContent}
      </Badge>
    )
  }
)

StatusBadge.displayName = "StatusBadge"

// Helper function to get status from goal state
export const getGoalStatus = (goal: {
  isCompleted?: boolean
  isPaused?: boolean
  isCancelled?: boolean
  dueDate?: Date | string | null
  startDate?: Date | string | null
}): keyof typeof statusLabels => {
  if (goal.isCancelled) return "cancelled"
  if (goal.isCompleted) return "completed"
  if (goal.isPaused) return "paused"
  
  const now = new Date()
  const dueDate = goal.dueDate ? new Date(goal.dueDate) : null
  const startDate = goal.startDate ? new Date(goal.startDate) : null
  
  if (dueDate && now > dueDate) return "overdue"
  if (startDate && now < startDate) return "pending"
  
  return "active"
}

// Helper function to get priority status
export const getPriorityStatus = (priority: string | number): keyof typeof statusLabels => {
  if (typeof priority === "string") {
    const lowerPriority = priority.toLowerCase()
    if (lowerPriority.includes("critical") || lowerPriority.includes("urgent")) return "critical"
    if (lowerPriority.includes("high")) return "high"
    if (lowerPriority.includes("medium") || lowerPriority.includes("normal")) return "medium"
    return "low"
  }
  
  // Numeric priority (1-4 scale)
  if (priority >= 4) return "critical"
  if (priority >= 3) return "high"
  if (priority >= 2) return "medium"
  return "low"
}

// Helper function to get progress status
export const getProgressStatus = (progress: number): keyof typeof statusLabels => {
  if (progress === 0) return "notStarted"
  if (progress >= 100) return "completed"
  return "inProgress"
}

export { StatusBadge, statusBadgeVariants, statusLabels, statusIcons }