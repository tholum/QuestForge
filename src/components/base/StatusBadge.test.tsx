import { render, screen } from '@testing-library/react'
import { describe, it, expect } from 'vitest'
import { StatusBadge, getGoalStatus, getPriorityStatus, getProgressStatus } from './StatusBadge'
import { Target } from 'lucide-react'

describe('StatusBadge', () => {
  it('renders with default status', () => {
    render(<StatusBadge status="active" />)
    expect(screen.getByText('Active')).toBeInTheDocument()
  })

  it('applies correct variant classes for different statuses', () => {
    const { rerender } = render(<StatusBadge status="completed" />)
    expect(screen.getByText('Completed')).toHaveClass('text-green-800')
    
    rerender(<StatusBadge status="error" />)
    expect(screen.getByText('Error')).toHaveClass('text-red-800')
    
    rerender(<StatusBadge status="warning" />)
    expect(screen.getByText('Warning')).toHaveClass('text-yellow-800')
  })

  it('shows custom label when provided', () => {
    render(<StatusBadge status="active" customLabel="In Progress" />)
    expect(screen.getByText('In Progress')).toBeInTheDocument()
  })

  it('shows count when provided', () => {
    render(<StatusBadge status="active" count={5} />)
    expect(screen.getByText('Active (5)')).toBeInTheDocument()
  })

  it('renders custom icon', () => {
    render(
      <StatusBadge 
        status="active" 
        customIcon={<Target data-testid="custom-icon" />}
      />
    )
    expect(screen.getByTestId('custom-icon')).toBeInTheDocument()
  })

  it('hides icon when showIcon is false', () => {
    render(<StatusBadge status="active" showIcon={false} />)
    
    // Should not have any svg elements (icons)
    const badge = screen.getByText('Active').closest('span')
    expect(badge?.querySelector('svg')).toBeNull()
  })

  it('applies size classes correctly', () => {
    const { rerender } = render(<StatusBadge status="active" size="sm" />)
    expect(screen.getByText('Active').closest('span')).toHaveClass('text-xs', 'px-2')
    
    rerender(<StatusBadge status="active" size="lg" />)
    expect(screen.getByText('Active').closest('span')).toHaveClass('px-3')
  })

  it('applies rounded classes correctly', () => {
    const { rerender } = render(<StatusBadge status="active" rounded="full" />)
    expect(screen.getByText('Active').closest('span')).toHaveClass('rounded-full')
    
    rerender(<StatusBadge status="active" rounded="none" />)
    expect(screen.getByText('Active').closest('span')).toHaveClass('rounded-none')
  })

  it('applies pulse animation when specified', () => {
    render(<StatusBadge status="active" pulse />)
    expect(screen.getByText('Active').closest('span')).toHaveClass('animate-pulse')
  })

  it('sets tooltip when provided', () => {
    render(<StatusBadge status="active" tooltip="This is active" />)
    expect(screen.getByText('Active').closest('span')).toHaveAttribute('title', 'This is active')
  })

  // Test automatic pulse for certain statuses
  it('automatically pulses for overdue status', () => {
    render(<StatusBadge status="overdue" />)
    expect(screen.getByText('Overdue').closest('span')).toHaveClass('animate-pulse')
  })

  it('automatically pulses for levelUp status', () => {
    render(<StatusBadge status="levelUp" />)
    expect(screen.getByText('Level Up!').closest('span')).toHaveClass('animate-pulse')
  })

  // Helper function tests
  describe('getGoalStatus', () => {
    it('returns completed for completed goals', () => {
      const goal = { isCompleted: true, isPaused: false, isCancelled: false }
      expect(getGoalStatus(goal)).toBe('completed')
    })

    it('returns cancelled for cancelled goals', () => {
      const goal = { isCompleted: false, isPaused: false, isCancelled: true }
      expect(getGoalStatus(goal)).toBe('cancelled')
    })

    it('returns paused for paused goals', () => {
      const goal = { isCompleted: false, isPaused: true, isCancelled: false }
      expect(getGoalStatus(goal)).toBe('paused')
    })

    it('returns overdue for overdue goals', () => {
      const yesterday = new Date(Date.now() - 86400000)
      const goal = { 
        isCompleted: false, 
        isPaused: false, 
        isCancelled: false, 
        dueDate: yesterday 
      }
      expect(getGoalStatus(goal)).toBe('overdue')
    })

    it('returns pending for future goals', () => {
      const tomorrow = new Date(Date.now() + 86400000)
      const goal = { 
        isCompleted: false, 
        isPaused: false, 
        isCancelled: false, 
        startDate: tomorrow 
      }
      expect(getGoalStatus(goal)).toBe('pending')
    })

    it('returns active for active goals', () => {
      const goal = { isCompleted: false, isPaused: false, isCancelled: false }
      expect(getGoalStatus(goal)).toBe('active')
    })
  })

  describe('getPriorityStatus', () => {
    it('correctly identifies string priorities', () => {
      expect(getPriorityStatus('low')).toBe('low')
      expect(getPriorityStatus('high')).toBe('high')
      expect(getPriorityStatus('critical')).toBe('critical')
      expect(getPriorityStatus('urgent')).toBe('critical')
    })

    it('correctly identifies numeric priorities', () => {
      expect(getPriorityStatus(1)).toBe('low')
      expect(getPriorityStatus(2)).toBe('medium')
      expect(getPriorityStatus(3)).toBe('high')
      expect(getPriorityStatus(4)).toBe('critical')
    })
  })

  describe('getProgressStatus', () => {
    it('returns notStarted for 0 progress', () => {
      expect(getProgressStatus(0)).toBe('notStarted')
    })

    it('returns inProgress for partial progress', () => {
      expect(getProgressStatus(45)).toBe('inProgress')
    })

    it('returns completed for 100% progress', () => {
      expect(getProgressStatus(100)).toBe('completed')
    })
  })

  // Accessibility tests
  it('has proper role and attributes', () => {
    render(<StatusBadge status="active" />)
    const badge = screen.getByText('Active').closest('span')
    
    // Badge should be properly labeled for screen readers
    expect(badge).toBeInTheDocument()
  })

  // Snapshot test
  it('matches snapshot for various configurations', () => {
    const { container } = render(
      <div>
        <StatusBadge status="active" />
        <StatusBadge status="completed" size="lg" />
        <StatusBadge status="overdue" count={3} pulse />
        <StatusBadge status="levelUp" customLabel="Achievement!" />
      </div>
    )
    expect(container).toMatchSnapshot()
  })
})