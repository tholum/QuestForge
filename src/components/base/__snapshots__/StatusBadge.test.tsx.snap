// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatusBadge > matches snapshot for various configurations 1`] = `
<div>
  <div>
    <span
      class="justify-center w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive overflow-hidden [a&]:hover:bg-accent [a&]:hover:text-accent-foreground inline-flex items-center gap-1.5 font-medium transition-all duration-200 bg-blue-100 text-blue-800 border-blue-200 dark:bg-blue-900/30 dark:text-blue-300 dark:border-blue-800 text-sm px-2.5 py-1 min-h-7 border rounded-md"
      data-slot="badge"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-play w-3 h-3"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 5a2 2 0 0 1 3.008-1.728l11.997 6.998a2 2 0 0 1 .003 3.458l-12 7A2 2 0 0 1 5 19z"
        />
      </svg>
      <span
        class="truncate"
      >
        Active
      </span>
    </span>
    <span
      class="justify-center w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive overflow-hidden [a&]:hover:bg-accent [a&]:hover:text-accent-foreground inline-flex items-center gap-1.5 font-medium transition-all duration-200 bg-green-100 text-green-800 border-green-200 dark:bg-green-900/30 dark:text-green-300 dark:border-green-800 text-sm px-3 py-1.5 min-h-8 border rounded-md"
      data-slot="badge"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-circle-check w-3 h-3"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="10"
        />
        <path
          d="m9 12 2 2 4-4"
        />
      </svg>
      <span
        class="truncate"
      >
        Completed
      </span>
    </span>
    <span
      class="justify-center w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive overflow-hidden [a&]:hover:bg-accent [a&]:hover:text-accent-foreground inline-flex items-center gap-1.5 font-medium transition-all duration-200 bg-red-100 text-red-800 border-red-200 dark:bg-red-900/30 dark:text-red-300 dark:border-red-800 text-sm px-2.5 py-1 min-h-7 border rounded-md animate-pulse"
      data-slot="badge"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-circle-alert w-3 h-3"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="10"
        />
        <line
          x1="12"
          x2="12"
          y1="8"
          y2="12"
        />
        <line
          x1="12"
          x2="12.01"
          y1="16"
          y2="16"
        />
      </svg>
      <span
        class="truncate"
      >
        Overdue
         (3)
      </span>
    </span>
    <span
      class="justify-center w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive overflow-hidden [a&]:hover:bg-accent [a&]:hover:text-accent-foreground inline-flex items-center gap-1.5 font-medium transition-all duration-200 bg-purple-100 text-purple-700 border-purple-200 dark:bg-purple-900/20 dark:text-purple-400 dark:border-purple-800 text-sm px-2.5 py-1 min-h-7 border rounded-md animate-pulse"
      data-slot="badge"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-zap w-3 h-3"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 14a1 1 0 0 1-.78-1.63l9.9-10.2a.5.5 0 0 1 .86.46l-1.92 6.02A1 1 0 0 0 13 10h7a1 1 0 0 1 .78 1.63l-9.9 10.2a.5.5 0 0 1-.86-.46l1.92-6.02A1 1 0 0 0 11 14z"
        />
      </svg>
      <span
        class="truncate"
      >
        Achievement!
      </span>
    </span>
  </div>
</div>
`;
